delta/imagery/imagery_dataset.py

```diff
@@ -253,42 +121,66 @@ def _tile_generator(self, i, is_labels): # pragma: no cover
-        preprocess = image.get_preprocess()
-        image.set_preprocess(None) # parallelize the preprocessing, not in disk i/o threadpool
-        bands = range(image.num_bands())
-
-        # read one row ahead of what we process now
-        next_buf = self._iopool.submit(lambda: image.read(tiles[0][0]))
-        for (c, (rect, sub_tiles)) in enumerate(tiles):
-            cur_buf = next_buf
-            if c + 1 < len(tiles):
-                # extra lambda to bind c in closure
-                next_buf = self._iopool.submit((lambda x: (lambda: image.read(tiles[x + 1][0])))(c))
-            if cur_buf is None:
-                continue
+        epoch = self._epoch[1 if is_labels else 0]
+        self._epoch[1 if is_labels else 0] += 1 # note: can only do by epoch now
+        images = [(self._labels if is_labels else self._images).load(i) for i in range(len(self._images))]
+        image_locks = {}
+        image_preprocesses = {}
+        for img in images:
+            image_locks[img] = threading.Lock()
+            image_preprocesses[img] = img.get_preprocess()
+            img.set_preprocess(None) # parallelize preprocessing outside lock
+
+        def tile_gen():
+            # use same seed for labels and not labels, differ by epoch
+            rand = random.Random(self._random_seed + epoch * 11617)
+            image_tiles = [(images[i], self._list_tiles(i)) for i in range(len(images))]
+            for (img, tiles) in image_tiles:
+                rand.shuffle(tiles)
+            image_tiles = [(img, iter(tiles)) for (img, tiles) in image_tiles]
+            while image_tiles:
+                index = rand.randrange(len(image_tiles))
+                (img, it) = image_tiles[index]
+                try:
+                    yield (img, next(it))
+                except StopIteration:
+                    del image_tiles[index]
+
+        def read_image(img, rect):
+            lock = image_locks[img]
+            preprocess = image_preprocesses[img]
+            lock.acquire()
+            buf = img.read(rect)
+            lock.release()
             if preprocess:
-                buf = preprocess(cur_buf.result(), rect, bands)
-            else:
-                buf = cur_buf.result()
-            (rect, sub_tiles) = tiles[c]
+                buf = preprocess(buf, rect, None)
+            return buf
+
+        def add_to_queue(buf_queue, item):
+            (img, (rect, sub_tiles)) = item
+            buf = self._iopool.submit(lambda: read_image(img, rect))
+            buf_queue.append((img, (rect, sub_tiles), buf))
+
+        gen = tile_gen()
+        buf_queue = []
+        for _ in range(config.io.threads() * 2): # add a bit ahead
+            try:
+                next_item = next(gen)
+            except StopIteration:
+                break
+            add_to_queue(buf_queue, next_item)
+        while buf_queue:
+            (img, (_, sub_tiles), buf) = buf_queue.pop(0)
```

Comment:
Why is just one of these outputs "blanked out" while only 'buf' is used?

Reply:
no, good point